Fix: #64. Combine headers. (#125)
ietf-wg-httpapi/ratelimit-headers: draft-ietf-httpapi-ratelimit-headers.md

--- abstract

> This document defines the RateLimit and RateLimit-Policy HTTP header fields for servers to advertise their current service rate limits, thereby allowing clients to avoid being throttled.

Reviewer: ```suggestion
This document defines the RateLimit-Policy and RateLimit HTTP header fields for servers to advertise their service policy limits and the current limits, thereby allowing clients to avoid being throttled.
```

Currently, there is no standard way for servers to communicate quotas so that clients can throttle its requests to prevent errors. This document defines a set of standard HTTP header fields to enable rate limiting:

- RateLimit: to convey

>   the server's quota for requests by the client in the time window,

Reviewer: ```suggestion
  the server's current limit of quota units available to the client in the policy time window,
```

>   the remaining quota in the current window,

Reviewer: ```suggestion
  the remaining quota units in the current window,
```

  and the time remaining in the current window, specified in seconds, and

> - RateLimit-Policy: the quota policy.

Reviewer: ```suggestion
- RateLimit-Policy: the service policy limits.
```

If they appear, they MUST be ignored.

The expiring limit MUST be set to the service limit that is closest to reaching its limit, and the associated time window MUST either be:

- inferred by the value of the [reset keyword](#ratelimit-reset-keyword) at the moment of the reset, or
- communicated out-of-band (e.g. in the documentation).

Example:

~~~ example
   limit=100
~~~

The RateLimit-Policy header field (see {{ratelimit-policy-field}}), might contain information on the associated time window.

## Remaining Keyword {#ratelimit-remaining-keyword}

The "remaining" keyword indicates the remaining quota units associated with the expiring-limit.

It is an Item and its value is a non-negative Integer expressed in [quota units](#service-limit).
This specification does not define Parameters for it.
If they appear, they MUST be ignored.

Clients MUST NOT assume that a positive remaining value is a guarantee that further requests will be served.

When the value of the remaining keyword is low, it indicates that the server may soon throttle the client (see {{providing-ratelimit-fields}}).

For example:

~~~ example
   remaining=50
~~~

## Reset Keyword {#ratelimit-reset-keyword}

> The "reset" keyword indicates the number of seconds until the quota associated with the expiring-limit resets.

Reviewer: ```suggestion
The "reset" keyword indicates the number of seconds until the available quota units associated with the expiring-limit resets.
```

It is a non-negative Integer compatible with the delay-seconds rule, because:

- it does not rely on clock synchronization and is resilient to clock adjustment
  and clock skew between client and server (see {{Section 5.6.7 of HTTP}});
- it mitigates the risk related to thundering herd when too many clients are serviced with the same timestamp.

This specification does not define Parameters for it.
If they appear, they MUST be ignored.

For example:

~~~ example
   reset=50
~~~

The client MUST NOT assume that all its service limit will be reset at the moment indicated by the reset keyword. The server MAY arbitrarily alter the reset keyword value between subsequent requests; for example, in case of resource saturation or to implement sliding window policies.

## RateLimit-Policy {#ratelimit-policy-field}

> The "RateLimit-Policy" response header field indicates the quota policies currently associated with the client. Its value is informative.

Reviewer: ```suggestion
The "RateLimit-Policy" response header field indicates a service policy currently associated with the client. Its value is informative.
```

The field is a non-empty List of Items. Each item is a [quota policy](#quota-policy).

> Two quota policies MUST NOT be associated with the same quota units value.

Reviewer: What is the reason for adding this constraint? And doesn't it contradict the example shown below with multiple time windows?
````
RateLimit-Policy: 10;w=1, 50;w=60, 1000;w=3600, 5000;w=86400
````
Is your assertion that each of these policies has a different quota unit?

Author: > Two quota policies MUST NOT be associated with the same quota units value.

I meant that the Policy field below is invalid

```
Policy: 10;w=1, 10;w=10
```
